### DISCUSSION TITLE
com.openai.unity 7.7.1

### DISCUSSION THREAD

**Discussion by StephenHodgson** [RageAgainstThePixel/com.openai.unity#197] [Announcements]
## What's Changed
- More Function utilities and invoking methods
  - Added FunctionPropertyAttribute to help better inform the feature how to format the Function json
  - Added FromFunc<,> overloads for convenance
  - Fixed invoke args sometimes being casting to wrong type
  - Added additional protections for static and instanced function calls
  - Added additional tool utilities:
    - Tool.ClearRegisteredTools
    - Tool.IsToolRegistered(Tool) - Tool.TryRegisterTool(Tool)
- Updated Sample Chat Behaviour with updated tool implementations
- com.utilities.rest -> 2.5.3

**Full Changelog**: https://github.com/RageAgainstThePixel/com.openai.unity/compare/7.7.0...7.7.1

<hr /><em>This discussion was created from the release <a href='https://github.com/RageAgainstThePixel/com.openai.unity/releases/tag/7.7.1'>com.openai.unity 7.7.1</a>.</em>

**Comment by pinkwerks** [comment]
Could someone elaborate on the `FunctionAttribute` a bit more? I read the README and it refers to finding methods with Function attributes, but I was hoping for an example of a method that used it. Previously I had a class where I had the schema with argument descriptions as well as the function itself. I was thrilled to see these attributes thinking it would be awesome to avoid all the JObject schema stuff, so I'm hoping to port code....
```csharp
public class GetPositionTool
{
    public class Args
    {
        public string ObjectName { get; set; }
    }

    public static OpenAI.Function FunctionSchema = new OpenAI.Function(
    nameof(GetPositionCmd),
    "Gets the position of an object",
    new JObject
    {
        ["type"] = "object",
        ["properties"] = new JObject
        {
            [nameof(Args.ObjectName)] = new JObject
            {
                ["type"] = "string",
                ["description"] = "Name of the object."
            },
        },
        ["required"] = new JArray { nameof(Args.ObjectName) }
    });

    internal static string GetPositionCmd(Args functionArgs)
    {
        var go = GameObject.Find(functionArgs.ObjectName);
        if (go != null)
        {
            return $"World space position: {go.transform.position}";
        }
        else
        {
            return $"Could't find object named {functionArgs.ObjectName}!";
        }
    }
}

--- (elsewhere in ChatBehavior.cs) ----
private readonly List<Tool> assistantTools = new List<Tool>();
assistantTools.Add(GetPositionTool.FunctionSchema);
var request = new ChatRequest(conversation.Messages, tools: assistantTools);

```
Now with 7.7 this is appearing easier, and at first glance I thought the Function attribute would make this much tidier... but I'm missing out on how to pass argument descriptions and if they are required. At first glance I was thinking I could refactor my code and write something like the below. 

Maybe another way of stating it: can `Tool.GetOrCreateTool` or `Tool.FromFunc` declare descriptions or optional parameters?

```csharp
[Function("Gets the position of an object")]
internal static string GetPositionCmd(
    [FunctionProperty("Name of the object.")]
    [Required]
    string ObjectName 
{
        ...
}

--- (elsewhere in ChatBehavior.cs) ---

var tools = Tool.GetAllAvailableTools();
var request = new ChatRequest(conversation.Messages, tools: tools);
```

**Reply by StephenHodgson** [reply]
Hi @pinkwerks! I added 3 new attributes in the 7.7.x releases so far:

Previously, you could define a function call as a json object, which was a manual process. Now you can specify tools by registering ***static*** methods, *instances* of objects and their method names, and by `func<,>` callbacks.

- **FuntionAttribute**: Marks ***static*** methods available for tool calls, and provides a way to optionally define a `description`.
- **FunctionParameterAttribute**: Adds a way to define a `description` to parameters of any method used in tool calls.
- **FunctionPropertyAttribute**: Adds a way to better define and refine the property json that is automatically generated by reflection. This includes a `description`, `required` flag, `defaultValue`, and `possibleValues`. This last attribute is really only useful if you have complex objects you're passing in as a parameter in your method.

**Reply by StephenHodgson** [reply]
On a side note, `[Required]` is not a valid parameter attribute for this system. Parameters are required by default unless you set a default value in the method like so:

```csharp
[Function("My Custom Function")]
public static void MyMethod(
    [FunctionParameter("A required value")] string requiredValue, 
    [FunctionParameter("An optional value")] string optionalValue = null) { }
```

**Reply by StephenHodgson** [reply]
Btw you can also register tools directly using `Tool.GetOrCreateTool(Type type, string methodName)`, `Tool.GetOrCreateTool(object instance, string methodName)` or `Tool.FromFunc(string name, Func<,> func)`

**Reply by StephenHodgson** [reply]
https://github.com/RageAgainstThePixel/com.openai.unity/blob/4402fa0747851cc219cca9dd2320c3fefc4295de/OpenAI/Packages/com.openai.unity/Tests/Weather/WeatherService.cs#L16-L19

**Comment by pinkwerks** [comment]
Wow this is so much nicer - Thanks for the extra explanations!

**Comment by pinkwerks** [comment]
Any hints on how to use arrays or lists as function parameters? I had Arrays working the old way, but it wasn't obvious how to migrate to the new stuff, I've tried `float[]` and `List<float>` with no luck.

I suppose I could just change the signature to have 3 floats in this particular case.... 🤔 You mentioned FunctionPropertyAttribute, but it didn't look like it supported Array types, or at least in a way obvious to me.

```
    public static string SetPosition(
        [FunctionParameter("X, Y, and Z position in world space")] float[] position)
```
Error: ArgumentException: Object of type 'Newtonsoft.Json.Linq.JArray' cannot be converted to type 'System.Single[]'.

**Reply by StephenHodgson** [reply]
Use Vector3 type in this case.

**Reply by pinkwerks** [reply]
Sorry forgot to mention I had tried a Vector3 but I get a bad request from the service:
```json
 {
          "type": "function",
          "function": {
            "name": "SetObjectPositionTool_SetObjectPositionCmd",
            "description": "Sets the position of an [object] in the scene.",
            "parameters": {
              "type": "object",
              "properties": {
                "objectName": {
                  "type": "string",
                  "description": "Name of the [object] to move"
                },
                "position": {
                  "type": "vector3",
                  "description": "X, Y, and Z position in world space"
                }
              },
              "required": [
                "objectName",
                "position"
              ]
            }
          }
        },
...

    "body": [
      {
        "error": {
          "message": "Invalid schema for function 'SetObjectPositionTool_SetObjectPositionCmd': 'vector3' is not valid under any of the given schemas.",
          "type": "invalid_request_error",
          "param": null,
          "code": null
        }
      }
    ],
    "error": "HTTP/1.1 400 Bad Request\n"
  }
```

**Reply by StephenHodgson** [reply]
use Vector3 type and let the GetOrCreateTool generate the schema for you. It'll be an `object` with 3 other properties that are `number` types.

**Reply by pinkwerks** [reply]
Sorry I must be doing something wrong, or not understanding your suggestion to let it create the schema for me. 

<img width="1592" alt="image" src="https://github.com/RageAgainstThePixel/com.openai.unity/assets/538256/b7f787c0-86c5-4997-84b6-cc76744b8243">

**Reply by pinkwerks** [reply]
Indeed, it looks like we are falling through to `schema["type"] = type.Name.ToLower();` in TypeExtention.cs 
 when seeing a Vector3, as type.IsClass is false. Which if true, I think would have produced the object like you suggested. But Vector3 seems to be a value type.
 https://github.com/RageAgainstThePixel/com.openai.unity/blob/43212ee6d0988ee7c5a4424a13a2d96e324a96af/OpenAI/Packages/com.openai.unity/Runtime/Extensions/TypeExtensions.cs#L85
 https://github.com/RageAgainstThePixel/com.openai.unity/blob/43212ee6d0988ee7c5a4424a13a2d96e324a96af/OpenAI/Packages/com.openai.unity/Runtime/Extensions/TypeExtensions.cs#L187

**Reply by StephenHodgson** [reply]
Looks like a bug! I'll try to sort that out in the next release then. Thanks

**Reply by StephenHodgson** [reply]
@pinkwerks try out https://github.com/RageAgainstThePixel/com.openai.unity/pull/205 and let me know if it solves the problem.

**Reply by pinkwerks** [reply]
Yes, Vector3 now seems to work - thanks for these updates! However with this function (and others) I'm seeing an exception I didn't see before.
```csharp
public class GetPositionTool
{
    [Function("Get the world space position of the object")]
    public static string GetPositionCmdNew(
        [FunctionParameter("Name of object to query")] string objectName)
    {
        var go = GameObject.Find(objectName);
        if (go != null)
        {
            var msg = $"{objectName} position: {go.transform.position}";
            Debug.Log(msg);
            return msg;
        }
        else
        {
            return $"Could't find object named {objectName}!";
        }
    }
}
```
![image](https://github.com/RageAgainstThePixel/com.openai.unity/assets/538256/6c7f343f-11fb-4900-ae94-df018f001720)

I am using `GetOrCreate` on that static method above. Do these static methods need to be Task types?
```csharp
        assistantTools.Add(
            Tool.GetOrCreateTool(
                typeof(GetPositionTool),
                nameof(GetPositionTool.GetPositionCmdNew)));
```

**Reply by StephenHodgson** [reply]
Can't read the screen shots. Plz put errors in text/code blocks.

**Reply by StephenHodgson** [reply]
You're getting invalid operation because the methods are not awaitable. Use regular Invoke, not InvokeAsync.

I'd need to see exactly where you're calling invoke on function call to know for sure.

**Reply by pinkwerks** [reply]
That's the thing, I'm not calling Invoke or InvokeAsync myself. The only code I've written is this (which worked ok before) 🤔
```
public class GetPositionTool
{
    [Function("Get the world space position of the object")]
    public static string GetPositionCmdNew(
        [FunctionParameter("Name of object to query")] string objectName)
    {
        var go = GameObject.Find(objectName);
        if (go != null)
        {
            var msg = $"{objectName} position: {go.transform.position}";
            Debug.Log(msg);
            return msg;
        }
        else
        {
            return $"Could't find object named {objectName}!";
        }
    }
}

--- (elsewhere in ChatBehaviour.cs) ---

 assistantTools.Add(
     Tool.GetOrCreateTool(
         typeof(GetPositionTool),
         nameof(GetPositionTool.GetPositionCmdNew)));

--- (stack trace) ---

InvalidOperationException: The function GetPositionTool_GetPositionCmdNew did not return a valid Task.
OpenAI.Function.InvokeAsync[T] (System.Threading.CancellationToken cancellationToken) (at ./Library/PackageCache/com.openai.unity@7.7.5/Runtime/Common/Function.cs:349)
UnityEngine.Debug:LogException(Exception)
OpenAI.<InvokeAsync>d__47`1:MoveNext() (at ./Library/PackageCache/com.openai.unity@7.7.5/Runtime/Common/Function.cs:360)
System.Runtime.CompilerServices.MoveNextRunner:Run()
Utilities.Async.AwaiterExtensions:RunOnUnityScheduler(Action) (at ./Library/PackageCache/com.utilities.async@2.1.3/Runtime/Async/AwaiterExtensions.cs:248)
Utilities.Async.SimpleCoroutineAwaiter:Complete(Exception) (at ./Library/PackageCache/com.utilities.async@2.1.3/Runtime/Async/SimpleCoroutineAwaiter.cs:47)
Utilities.Async.<ReturnVoid>d__27:MoveNext() (at ./Library/PackageCache/com.utilities.async@2.1.3/Runtime/Async/AwaiterExtensions.cs:334)
UnityEngine.SetupCoroutine:InvokeMoveNext(IEnumerator, IntPtr)
```
`Function.cs` is doing a `function.MethodInfo.Invoke()` afaik. Should that "is not" be "is"?

https://github.com/RageAgainstThePixel/com.openai.unity/blob/052512c93cc2dae335b40a5cba9a927654fba0e4/OpenAI/Packages/com.openai.unity/Runtime/Common/Function.cs#L347

**Reply by StephenHodgson** [reply]
You got to be invoking it somewhere.
Or else you wouldn't get to this point.

Where are you stopping to handle the tool outputs?

**Reply by pinkwerks** [reply]
Ah my bad, I was relying heavily on the Sample's ChatBehavior.cs which calls `InvokeFunctionAsync` as part of Submit chat. When I updated the package and sample, I forgot I had patched up that function to deal with more tools than the non-async image generation stuff. Thanks again!
